File: go/internal/transport.go
package internal

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"math"
	"net/http"
	"strconv"
	"strings"
	"time"

	"github.com/ai-aegis-yatav/aegis-sdk/go/apierrors"
	"github.com/ai-aegis-yatav/aegis-sdk/go/models"
	"github.com/ai-aegis-yatav/aegis-sdk/go/streaming"
)

const (
	SDKVersion = "0.1.0"
	UserAgent  = "aegis-go-sdk/" + SDKVersion
)

type Transport struct {
	HTTPClient    *http.Client
	BaseURL       string
	APIKey        string
	MaxRetries    int
	CustomHeaders map[string]string
	LastQuota     *models.QuotaInfo
}

func NewTransport(apiKey, baseURL string, timeout time.Duration, maxRetries int, headers map[string]string) *Transport {
	return &Transport{
		HTTPClient:    &http.Client{Timeout: timeout},
		BaseURL:       strings.TrimRight(baseURL, "/"),
		APIKey:        apiKey,
		MaxRetries:    maxRetries,
		CustomHeaders: headers,
	}
}

// Do executes an HTTP request with retry logic, parsing the JSON response into result.
func (t *Transport) Do(ctx context.Context, method, path string, body interface{}, result interface{}) error {
	var reqBody []byte
	if body != nil {
		var err error
		reqBody, err = json.Marshal(body)
		if err != nil {
			return &apierrors.AegisError{Message: "failed to marshal request body", Cause: err}
		}
	}

	var lastErr error
	for attempt := 0; attempt <= t.MaxRetries; attempt++ {
		if attempt > 0 {
			wait := t.backoff(attempt)
			select {
			case <-ctx.Done():
				return ctx.Err()
			case <-time.After(wait):
			}
		}

		var bodyReader io.Reader
		if reqBody != nil {
			bodyReader = bytes.NewReader(reqBody)
		}

		req, err := http.NewRequestWithContext(ctx, method, t.BaseURL+path, bodyReader)
		if err != nil {
			return &apierrors.NetworkError{AegisError: apierrors.AegisError{Message: "failed to create request", Cause: err}}
		}
		t.setHeaders(req)

		resp, err := t.HTTPClient.Do(req)
		if err != nil {
			lastErr = &apierrors.NetworkError{AegisError: apierrors.AegisError{Message: "request failed", Cause: err}}
			continue
		}

		t.parseQuotaHeaders(resp)

		if resp.StatusCode >= 200 && resp.StatusCode < 300 {
			if result != nil {
				defer resp.Body.Close()
				if err := json.NewDecoder(resp.Body).Decode(result); err != nil {
					return &apierrors.AegisError{Message: "failed to decode response", Cause: err}
				}
			} else {
				resp.Body.Close()
			}
			return nil
		}

		apiErr := t.parseError(resp)
		resp.Body.Close()

		if t.shouldRetry(resp.StatusCode) && attempt < t.MaxRetries {
			lastErr = apiErr
			continue
		}
		return apiErr
	}
	return lastErr
}

// DoStream executes an HTTP request and returns a StreamReader for SSE events.
func (t *Transport) DoStream(ctx context.Context, method, path string, body interface{}) (*streaming.StreamReader, error) {
	var bodyReader io.Reader
	if body != nil {
		data, err := json.Marshal(body)
		if err != nil {
			return nil, &apierrors.AegisError{Message: "failed to marshal request body", Cause: err}
		}
		bodyReader = bytes.NewReader(data)
	}

	req, err := http.NewRequestWithContext(ctx, method, t.BaseURL+path, bodyReader)
	if err != nil {
		return nil, &apierrors.NetworkError{AegisError: apierrors.AegisError{Message: "failed to create request", Cause: err}}
	}
	t.setHeaders(req)
	req.Header.Set("Accept", "text/event-stream")

	resp, err := t.HTTPClient.Do(req)
	if err != nil {
		return nil, &apierrors.NetworkError{AegisError: apierrors.AegisError{Message: "stream request failed", Cause: err}}
	}

	if resp.StatusCode >= 200 && resp.StatusCode < 300 {
		t.parseQuotaHeaders(resp)
		return streaming.NewStreamReader(resp.Body), nil
	}

	apiErr := t.parseError(resp)
	resp.Body.Close()
	return nil, apiErr
}

func (t *Transport) setHeaders(req *http.Request) {
	req.Header.Set("X-API-Key", t.APIKey)
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("User-Agent", UserAgent)
	for k, v := range t.CustomHeaders {
		req.Header.Set(k, v)
	}
}

func (t *Transport) parseQuotaHeaders(resp *http.Response) {
	limit := resp.Header.Get("X-Quota-Limit")
	used := resp.Header.Get("X-Quota-Used")
	remaining := resp.Header.Get("X-Quota-Remaining")
	resetAt := resp.Header.Get("X-Quota-Reset")
	tier := resp.Header.Get("X-Quota-Tier")

	if limit == "" && used == "" {
		return
	}

	q := &models.QuotaInfo{Tier: tier}
	if v, err := strconv.Atoi(limit); err == nil {
		q.Limit = v
	}
	if v, err := strconv.Atoi(used); err == nil {
		q.Used = v
	}
	if v, err := strconv.Atoi(remaining); err == nil {
		q.Remaining = v
	}
	if ts, err := time.Parse(time.RFC3339, resetAt); err == nil {
		q.ResetAt = ts
	}
	t.LastQuota = q
}

func (t *Transport) parseError(resp *http.Response) error {
	bodyBytes, _ := io.ReadAll(resp.Body)
	bodyStr := string(bodyBytes)

	var errResp struct {
		Message      string            `json:"message"`
		ErrorCode    string            `json:"error_code"`
		RequestID    string            `json:"request_id"`
		RequiredTier string            `json:"required_tier"`
		UpgradeURL   string            `json:"upgrade_url"`
		Limit        int               `json:"limit"`
		Used         int               `json:"used"`
		ResetAt      string            `json:"reset_at"`
		RetryAfter   float64           `json:"retry_after"`
		Fields       map[string]string `json:"fields"`
	}
	json.Unmarshal(bodyBytes, &errResp)

	base := apierrors.ApiError{
		StatusCode: resp.StatusCode,
		ErrorCode:  errResp.ErrorCode,
		RequestID:  errResp.RequestID,
		Body:       bodyStr,
		Message:    errResp.Message,
	}
	if base.Message == "" {
		base.Message = fmt.Sprintf("HTTP %d", resp.StatusCode)
	}

	switch resp.StatusCode {
	case 401:
		return &apierrors.AuthenticationError{ApiError: base}
	case 403:
		return &apierrors.TierAccessError{
			ApiError:     base,
			RequiredTier: errResp.RequiredTier,
			UpgradeURL:   errResp.UpgradeURL,
		}
	case 404:
		return &apierrors.NotFoundError{ApiError: base}
	case 422:
		return &apierrors.ValidationError{ApiError: base, Fields: errResp.Fields}
	case 429:
		retryAfter := time.Duration(errResp.RetryAfter) * time.Second
		if retryAfter == 0 {
			if ra := resp.Header.Get("Retry-After"); ra != "" {
				if secs, err := strconv.Atoi(ra); err == nil {
					retryAfter = time.Duration(secs) * time.Second
				}
			}
		}
		if errResp.Limit > 0 {
			resetTime, _ := time.Parse(time.RFC3339, errResp.ResetAt)
			return &apierrors.QuotaExceededError{
				ApiError: base,
				Limit:    errResp.Limit,
				Used:     errResp.Used,
				ResetAt:  resetTime,
			}
		}
		return &apierrors.RateLimitError{ApiError: base, RetryAfter: retryAfter}
	default:
		if resp.StatusCode >= 500 {
			return &apierrors.ServerError{ApiError: base}
		}
		return &apierrors.ApiError{
			StatusCode: base.StatusCode,
			ErrorCode:  base.ErrorCode,
			RequestID:  base.RequestID,
			Body:       base.Body,
			Message:    base.Message,
		}
	}
}

func (t *Transport) shouldRetry(statusCode int) bool {
	return statusCode == 429 || statusCode >= 500
}

func (t *Transport) backoff(attempt int) time.Duration {
	base := 500 * time.Millisecond
	max := 30 * time.Second
	wait := time.Duration(float64(base) * math.Pow(2, float64(attempt-1)))
	if wait > max {
		wait = max
	}
	return wait
}

File: python/src/aegis/_transport.py
"""HTTP transport layer with retry, error handling, and quota tracking."""

from __future__ import annotations

import time
from dataclasses import dataclass, field
from typing import Any, Dict, Optional

import httpx

from aegis.config import ClientConfig
from aegis.errors import (
    AuthenticationError,
    NetworkError,
    NotFoundError,
    QuotaExceededError,
    RateLimitError,
    ServerError,
    TierAccessError,
    ValidationError,
)


@dataclass
class QuotaInfo:
    """Tracks API quota usage from response headers."""

    limit: Optional[int] = None
    used: Optional[int] = None
    remaining: Optional[int] = None

    @property
    def is_available(self) -> bool:
        return self.limit is not None


@dataclass
class ApiResponse:
    """Wraps an HTTP response with parsed metadata."""

    status_code: int
    data: Any
    headers: Dict[str, str] = field(default_factory=dict)
    request_id: Optional[str] = None
    quota: Optional[QuotaInfo] = None


def _parse_quota(headers: httpx.Headers) -> QuotaInfo:
    info = QuotaInfo()
    if "x-quota-limit" in headers:
        try:
            info.limit = int(headers["x-quota-limit"])
        except ValueError:
            pass
    if "x-quota-used" in headers:
        try:
            info.used = int(headers["x-quota-used"])
        except ValueError:
            pass
    if "x-quota-remaining" in headers:
        try:
            info.remaining = int(headers["x-quota-remaining"])
        except ValueError:
            pass
    return info


def _raise_for_status(response: httpx.Response, request_id: Optional[str]) -> None:
    if response.status_code < 400:
        return

    try:
        body = response.json()
    except Exception:
        body = {"error": response.text}

    message = body.get("error", body.get("message", f"HTTP {response.status_code}"))
    error_code = body.get("code")
    kwargs: Dict[str, Any] = {
        "error_code": error_code,
        "request_id": request_id,
        "body": body,
    }

    status = response.status_code

    if status == 401:
        raise AuthenticationError(message, **kwargs)

    if status == 403:
        raise TierAccessError(
            message,
            required_tier=body.get("required_tier"),
            upgrade_url=body.get("upgrade_url"),
            **kwargs,
        )

    if status == 404:
        raise NotFoundError(message, **kwargs)

    if status == 429:
        retry_after = response.headers.get("retry-after")
        if body.get("code") == "QUOTA_EXCEEDED" or "quota" in message.lower():
            raise QuotaExceededError(
                message,
                limit=body.get("limit"),
                used=body.get("used"),
                reset_at=body.get("reset_at"),
                **kwargs,
            )
        raise RateLimitError(
            message,
            retry_after=float(retry_after) if retry_after else None,
            **kwargs,
        )

    if status in (400, 422):
        raise ValidationError(message, status_code=status, **kwargs)

    if status >= 500:
        raise ServerError(message, status_code=status, **kwargs)


class SyncTransport:
    """Synchronous HTTP transport with retry and exponential backoff."""

    def __init__(self, config: ClientConfig) -> None:
        self._config = config
        self.quota = QuotaInfo()
        self._client = httpx.Client(
            base_url=config.base_url,
            timeout=config.timeout,
            verify=config.verify_ssl,
            headers=self._build_headers(),
        )

    def _build_headers(self) -> Dict[str, str]:
        headers: Dict[str, str] = {
            "X-API-Key": self._config.api_key,
            "Content-Type": "application/json",
            "Accept": "application/json",
            "User-Agent": "aegis-python-sdk/0.1.0",
        }
        headers.update(self._config.custom_headers)
        return headers

    def request(
        self,
        method: str,
        path: str,
        *,
        json: Optional[Any] = None,
        params: Optional[Dict[str, Any]] = None,
        headers: Optional[Dict[str, str]] = None,
    ) -> ApiResponse:
        last_exc: Optional[Exception] = None

        for attempt in range(self._config.max_retries + 1):
            try:
                response = self._client.request(
                    method,
                    path,
                    json=json,
                    params=params,
                    headers=headers,
                )
                request_id = response.headers.get("x-request-id")
                self.quota = _parse_quota(response.headers)

                if response.status_code == 429 and attempt < self._config.max_retries:
                    retry_after = response.headers.get("retry-after")
                    wait = float(retry_after) if retry_after else self._backoff(attempt)
                    time.sleep(wait)
                    continue

                if response.status_code >= 500 and attempt < self._config.max_retries:
                    time.sleep(self._backoff(attempt))
                    continue

                _raise_for_status(response, request_id)

                data = response.json() if response.content else None
                return ApiResponse(
                    status_code=response.status_code,
                    data=data,
                    headers=dict(response.headers),
                    request_id=request_id,
                    quota=self.quota,
                )

            except (httpx.ConnectError, httpx.TimeoutException) as exc:
                last_exc = exc
                if attempt < self._config.max_retries:
                    time.sleep(self._backoff(attempt))
                    continue
                raise NetworkError(str(exc), cause=exc) from exc

        raise NetworkError("Max retries exceeded", cause=last_exc)

    def stream(
        self,
        method: str,
        path: str,
        *,
        json: Optional[Any] = None,
        params: Optional[Dict[str, Any]] = None,
    ) -> httpx.Response:
        headers = {"Accept": "text/event-stream"}
        return self._client.stream(
            method,
            path,
            json=json,
            params=params,
            headers=headers,
        )

    def _backoff(self, attempt: int) -> float:
        return self._config.retry_backoff_factor * (2**attempt)

    def close(self) -> None:
        self._client.close()


class AsyncTransport:
    """Asynchronous HTTP transport with retry and exponential backoff."""

    def __init__(self, config: ClientConfig) -> None:
        self._config = config
        self.quota = QuotaInfo()
        self._client = httpx.AsyncClient(
            base_url=config.base_url,
            timeout=config.timeout,
            verify=config.verify_ssl,
            headers=self._build_headers(),
        )

    def _build_headers(self) -> Dict[str, str]:
        headers: Dict[str, str] = {
            "X-API-Key": self._config.api_key,
            "Content-Type": "application/json",
            "Accept": "application/json",
            "User-Agent": "aegis-python-sdk/0.1.0",
        }
        headers.update(self._config.custom_headers)
        return headers

    async def request(
        self,
        method: str,
        path: str,
        *,
        json: Optional[Any] = None,
        params: Optional[Dict[str, Any]] = None,
        headers: Optional[Dict[str, str]] = None,
    ) -> ApiResponse:
        import asyncio

        last_exc: Optional[Exception] = None

        for attempt in range(self._config.max_retries + 1):
            try:
                response = await self._client.request(
                    method,
                    path,
                    json=json,
                    params=params,
                    headers=headers,
                )
                request_id = response.headers.get("x-request-id")
                self.quota = _parse_quota(response.headers)

                if response.status_code == 429 and attempt < self._config.max_retries:
                    retry_after = response.headers.get("retry-after")
                    wait = float(retry_after) if retry_after else self._backoff(attempt)
                    await asyncio.sleep(wait)
                    continue

                if response.status_code >= 500 and attempt < self._config.max_retries:
                    await asyncio.sleep(self._backoff(attempt))
                    continue

                _raise_for_status(response, request_id)

                data = response.json() if response.content else None
                return ApiResponse(
                    status_code=response.status_code,
                    data=data,
                    headers=dict(response.headers),
                    request_id=request_id,
                    quota=self.quota,
                )

            except (httpx.ConnectError, httpx.TimeoutException) as exc:
                last_exc = exc
                if attempt < self._config.max_retries:
                    await asyncio.sleep(self._backoff(attempt))
                    continue
                raise NetworkError(str(exc), cause=exc) from exc

        raise NetworkError("Max retries exceeded", cause=last_exc)

    async def stream(
        self,
        method: str,
        path: str,
        *,
        json: Optional[Any] = None,
        params: Optional[Dict[str, Any]] = None,
    ) -> Any:
        headers = {"Accept": "text/event-stream"}
        return self._client.stream(
            method,
            path,
            json=json,
            params=params,
            headers=headers,
        )

    def _backoff(self, attempt: int) -> float:
        return self._config.retry_backoff_factor * (2**attempt)

    async def close(self) -> None:
        await self._client.aclose()
